neurons/validator.py

```diff
@@ -651,88 +388,75 @@ def set_weights(self):
-                    )
-        # Update the iterator. It will keep its current position if possible.
-        self.miner_iterator.set_miner_uids(self.get_miner_uids(self.metagraph))
-
-        # Check to see if the metagraph has changed size.
-        # If so, we need to add new hotkeys and moving averages.
-        if len(self.hotkeys) < len(self.metagraph.hotkeys):
-            self.scorer.resize(self.metagraph.n)
-
-        # Update the hotkeys.
-        self.hotkeys = copy.deepcopy(self.metagraph.hotkeys)
+    @property
+    def block(self):
+        return utils.ttl_get_block(self)
 
     def save_state(self):
         """Saves the state of the validator to a file."""
         bt.logging.trace("Saving validator state.")
 
-        if not os.path.exists(self.config.neuron.full_path):
-            os.makedirs(self.config.neuron.full_path)
-
-        # Save the state of the validator to file.
-        self.scorer.save_state(
-            os.path.join(self.config.neuron.full_path, Validator.SCORER_FILENAME)
-        )
+        self.evaluator.save_state()
 
     def load_state(self):
         """Loads the state of the validator from a file."""
         bt.logging.info("Loading validator state.")
 
-        # Load the state of the validator from file.
-        filepath = os.path.join(self.config.neuron.full_path, Validator.SCORER_FILENAME)
-        if not os.path.exists(filepath):
-            bt.logging.warning("No scorer state file found. Starting from scratch.")
-            return
+        self.evaluator.load_state()
 
-        self.scorer.load_state(filepath)
-        bt.logging.success(f"Loaded scorer state from: {filepath}.")
+    def check_registered(self):
+        # --- Check for registration.
+        if self.wallet.hotkey.ss58_address not in self.metagraph.hotkeys:
+            bt.logging.error(
+                f"Wallet: {self.wallet} is not registered on netuid {self.config.netuid}."
+                f" Please register the hotkey using `btcli subnets register` before trying again."
+            )
+            sys.exit(1)
 
-        # Resize the scorer in case the loaded state is old and missing newly added neurons.
-        self.scorer.resize(self.metagraph.n)
 
+def main():
+    """Main constructs the validator with its dependencies."""
 
-# The main function parses the configuration and runs the validator.
-if __name__ == "__main__":
-    with Validator() as validator:
+    config = create_config(NeuronType.VALIDATOR)
+    subtensor = bt.subtensor(config=config)
+    metagraph = subtensor.metagraph(netuid=config.netuid)
+    wallet = bt.wallet(config=config)
+
+    # First, check the wallet is registered.
+    uid = 0
+    if wallet.hotkey.ss58_address in metagraph.hotkeys:
+        uid = metagraph.hotkeys.index(wallet.hotkey.ss58_address)
+        bt.logging.info(
+            f"Running neuron on subnet: {config.netuid} with uid {uid} using network: {subtensor.chain_endpoint}."
+        )
+    else:
+        bt.logging.warning(
```

Comment:
maybe a todo that once we have offline mode this could just be a check_registered call.

Reply:
Vali won't run in offline mode but I've moved this functionality to utils

Reply:
Then shouldn't this fail to start the vali? If there isn't a case where this isn't in the metagraph? Or is that what happens when you run on testnet?

Reply:
Good point.